runtime/jailer.go

```diff
@@ -0,0 +1,299 @@
+			if err := os.MkdirAll(filepath.Join(devPath, "net"), 0777); err != nil {
+				return errors.Wrapf(err, "failed to create device path: %v", devPath)
+			}
+
+			contentsPath := filepath.Join(rootPath, "var", "lib", "firecracker-containerd")
+			logger.Debugf("Creating firecracker contents path %v", contentsPath)
+			if err := os.MkdirAll(contentsPath, 0777); err != nil {
+				return errors.Wrapf(err, "failed to create contents path: %v", contentsPath)
+			}
+
+			rootPathToConfig := filepath.Join(jailPath, "config.json")
+			logger.Debugf("Copying config to %v", rootPathToConfig)
+			if err := copyFile(runcConfigPath, rootPathToConfig); err != nil {
+				return errors.Wrapf(err, "failed to copy config from %v to %v", runcConfigPath, rootPathToConfig)
+			}
+
+			logger.Debugf("Overwritting process args of config")
+			if err := overwriteConfig(cfg, filepath.Base(m.Cfg.SocketPath), rootPathToConfig); err != nil {
+				return errors.Wrap(err, "failed to overwrite config.json")
+			}
+
+			logger.Debugf("chdir to %v", jailPath)
+			if err := os.Chdir(jailPath); err != nil {
+				return errors.Wrapf(err, "failed to chdir into %v", jailPath)
+			}
+
+			// copy the firecracker binary
+			logger.Debugf("copying firecracker binary to %v", binPath)
+			if err := copyFile(cfg.FirecrackerBinaryPath, binPath); err != nil {
+				return errors.Wrapf(err, "could not copy firecracker binary from path %v", cfg.FirecrackerBinaryPath)
+			}
+
+			// Link the kernel image
+			newKernelImagePath := filepath.Join(contentsPath, kernelImageFileName)
+			logger.Debugf("linking kernel image to %v", newKernelImagePath)
+			if err := os.Link(m.Cfg.KernelImagePath, newKernelImagePath); err != nil {
+				return errors.Wrapf(err, "failed to link kernel image from path %v", m.Cfg.KernelImagePath)
+			}
+			m.Cfg.KernelImagePath = filepath.Join("/var/lib/firecracker-containerd", kernelImageFileName)
+
+			logger.Debugf("creating devices in %v", devPath)
+			if err := createDevices(devPath); err != nil {
+				return errors.Wrap(err, "failed to create necessary devices")
+			}
+
+			// link drives to new contents path
+			for i, d := range m.Cfg.Drives {
+				drivePath := firecracker.StringValue(d.PathOnHost)
+				fileName := filepath.Base(drivePath)
+
+				if err := os.Link(drivePath, filepath.Join(contentsPath, fileName)); err != nil {
+					return errors.Wrapf(err, "failed to copy drive %v", drivePath)
+				}
+
+				newDrivePath := filepath.Join(strings.TrimPrefix(contentsPath, rootPath), fileName)
+				m.Cfg.Drives[i].PathOnHost = firecracker.String(newDrivePath)
+			}
+
+			if err := chownr(rootPath, firecracker.IntValue(cfg.UID), firecracker.IntValue(cfg.GID)); err != nil {
+				return errors.Wrapf(err, "failed to chown -r %v", rootPath)
+			}
+
+			logger.Info("Successfully ran jailer handler")
+			return nil
+		},
+	}
+}
+
+// createDevices will create a series of device nodes at the given path
+func createDevices(path string) error {
```

Comment:
Did you give a try to doing this with the runc config directly? You can specify devices that should be created (including major/minor numbers). Search for `/dev/fuse` on this page to see an example: https://github.com/opencontainers/runtime-spec/blob/master/config.md

If you are able to do that, I would also give a try to adding `/dev` as a tmpfs mount in the runc config directly, which is normally what's done.

Reply:
Yea, I did try that but it was behaving a little finicky. My thinking was since this is absolutely needed by Firecracker, it doesn't hurt to always create them and not rely on the config in case someone forgets.

Reply:
What was finicky about it? It should work consistently. I think it's worth doing too because it allows us to remove code here that creates the directories for the devices and makes the devices themselves (runc already has all the code for doing that).

As for users forgetting about them, I agree that is a valid concern but another way to deal with that and similar situations is by validating the config that is being provided by the users, which I think would be good in that it's a more general mechanism in addition to the above mentioned benefit.